pkg/pillar/cmd/domainmgr/domainmgr.go

```diff
@@ -1116,6 +1149,154 @@ func lookupDomainConfig(ctx *domainContext, key string) *types.DomainConfig {
 	return &config
 }
 
+func runningOnKVM() bool {
+	_, enabledHVs := hypervisor.GetAvailableHypervisors()
+	for _, enabledHV := range enabledHVs {
+		if enabledHV == hypervisor.KVMHypervisorName {
+			return true
+		}
+	}
+	return false
+}
+
+func setCgroupCpuset(vmUUIDAndVersion string, cpus string) error {
+	if !runningOnKVM() {
+		log.Warnf("running not on KVM/QEMU system, skip setting croups")
+		return nil
+	}
+	cpuSetFileName := fmt.Sprintf("/sys/fs/cgroup/cpuset/eve-user-apps/%s/cpuset.cpus", vmUUIDAndVersion)
+	subCommand := fmt.Sprintf("echo %s > %s", cpus, cpuSetFileName)
+	// Has to run with access to host FS. The corresponding cpuset directory is not accessible from within a container
+	err := exec.Command("chroot", "/hostfs", "sh", "-c", subCommand).Run()
+	if err != nil {
+		log.Warnf("Failed to execute: 'chroot /hostfs sh -c %s'", subCommand)
+		return err
+	}
+	return nil
+}
+
+func constructNonPinnedCpumaskString(ctx *domainContext) string {
+	result := ""
+	for cpu, pinned := range ctx.hostCpusPinned {
+		if !pinned {
+			addToMask(cpu, &result)
+		}
+	}
+	return result
+}
+
+func addToMask(cpu int, s *string) {
+	if s == nil {
+		return
+	}
+	if *s == "" {
+		*s = fmt.Sprintf("%d", cpu)
+	} else {
+		*s = fmt.Sprintf("%s,%d", *s, cpu)
+	}
+}
+
+func excludeCPUFromNonPinnedVMs(ctx *domainContext, cpu int) error {
+	log.Functionf("Pinning a CPU (%d), exclude it from the other VMs CPU sets...", cpu)
+	ctx.hostCpusPinned[cpu] = true
+	//Exclude the CPU from the shared list and fix the cgroups of previously created VMs
+	for vmUUID := range ctx.resourcesUsedByVMs {
+		vmToCheck := ctx.resourcesUsedByVMs[vmUUID]
+		// Need the Status here only to get the proper Key to find the corresponding Config
+		vmStatus := lookupDomainStatusByUUID(ctx, vmUUID)
+		if vmStatus == nil {
+			//Most probably the status has not been created yet
+			continue
+		}
+		if vmStatus.VmConfig.CPUsPinned {
+			// We need to check only non-pinned VMs
+			continue
+		}
+		vmConfig := lookupDomainConfig(ctx, vmStatus.Key())
+		if vmConfig == nil {
+			// Strange situation. Should be investigated if happens.
+			log.Warnf("Failed to find the DomainConfig corresponding to a DomainStatus (%s)", vmStatus.DisplayName)
```

Comment:
I'm not ceertain, but I think we can have a DomainStatus without a DomainConfig when the controller has deleted the app instance, and domainmgr is waiting for the domain/task to exit.
Isn't there code in domainmgr to handle such a state elsewhere?

Reply:
Mmm... Should I do anything here?
@giggsoff @eriknordmark

Reply:
Sorry, seems the code is outdated, and I cannot remeber the context.
But as I can see in case of delete of DomainConfig we will go into handleDelete and will run [doInactivate](https://github.com/lf-edge/eve/blob/1844bc4b5a825fd33a06b7630136a6cbc6e3797d/pkg/pillar/cmd/domainmgr/domainmgr.go#L2149) only if status.Activated set. If status.Activated not set (for example we stop the domain and decide to delete it), we will not call `ctx.cpuAllocator.Free` function.

Reply:
But if a Domain is stopped, it means, that it was inactivated before? So, the Free function should have been called by that moment? Or did I miss something?

Reply:
You can have stopped domain. After reboot you will do cpu assignment in handleCreate. But if you will decide to delete the domain, you will not call doInactivate as domain was stopped.

Reply:
Hmmm... That's interesting. It means that if we Create a VM -> Stop It -> Reboot the board - the VM will occupy the CPUs. I should check it, but I think I had this use case and it looks like the VM didn't take the CPUs. Let me check.
Thanks for a good corner case!

Reply:
I'm still testing. But according to the code, I think @giggsoff , you are right. 
What would be the solution? To put the `assignCPUs` call in the `handleCreate` function under the `if config.Activate {` condition?

Reply:
If you want to allocate cpus for the whole life cycle of edge-node regardless of activation state of domain (only depends on existence of domain), I would recomend to move free into handleDelete.
If you want allocate cpus and free them regarding the state of domain (allocate on start, de-allocate pn stop), you can move allocation into doActivate.
It is better to keep symmetry in (de-)allocators: handleCreate/handleDelete or doActivate/doInactivate.

Reply:
Hmmm... I think @eriknordmark wanted to see the CPUs deallocation in `doInactivate()`... So, I think I will try to move the allocation to `doActivate()` now.

Reply:
That's strange, but I cannot reproduce the situation. If I create a VM, stop it and then reboot - the CPUs are not occupied by the stopped VM. 
Nevertheless, I'm going to mode the allocation to doActivate(), just for symmetry.

Reply:
You mean, that we do not enter handleCreate function in that case (it is strange)? Or stopped domain has no effect to other domains (probably we do not trigger cpu notification in time, but it may be triggered by the third domain eventually)?

Reply:
I do not see that the CPUs that should be allocated to the stopped VM are gone from the free list of CPUs, that are assigned to the VMs with no CPU pinning. 
Nevertheless, I moved the allocation code. Tested it. Will push now.

Reply:
Done. May I resolve the thread?

Reply:
Thank you, it looks like the correct and symmetry place now. But sorry, not sure about the thread as Erik started it.

Reply:
Ok. Consider it as done.